### ISSUE TITLE
SIMH networking

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#182]
Write a driver for the DEUNA/DELUA Ethernet controller supported by SIMH.

**Comment by larsbrinkhoff** [comment]
AI: JTW; KSTEST > is a standalone test tool.  I has support for testing an Interlan NI1010A ethernet interface.

I'm not sure whether ITS supports the interface.

**Comment by markpizz** [comment]
Someone could also take a shot at implementing a simh device simulator for the Interlan NI1010A...

**Comment by larsbrinkhoff** [comment]
Yes, but it would be a gamble.  We don't know if ITS can use the NI1010A.  I should probably contact the person who was writing the driver code.

**Comment by larsbrinkhoff** [comment]
@jtw43, did you finish the NI1010A driver?

**Comment by larsbrinkhoff** [comment]
Default SIMH DELUA configuration, as displayed by `show xy`:

    XU    address=3774510-3774517, vector=120, MAC=08:00:2B:BC:CE:48
    type=DELUA, throttle=disabled
    not attached

It would be intersting to poke at it and see what happens.

**Comment by larsbrinkhoff** [comment]
I have made an .IOTLSR programs which sends a few commands to the device, and triggers an interrupt.

**Comment by larsbrinkhoff** [comment]
Adding a `UBIDV 120` to the KS10 UNIBUS page in ITS ensures that the interrupt is handled.  Probably by the UBI: device which may not be that useful.

May need a `UBINT 120, JSR DEUBK` to get control over the interrupt.

**Comment by larsbrinkhoff** [comment]
I have now received word from @jtw43 that the NI1010 work indeed was not finished:
> My memory about the unibus ethernet adapters is fuzzier. IIRC we ran into serious problems in the end with the 18/16 bit unibus data formatting and byte swapping issues, and decided that performance would be not worth it.

**Comment by larsbrinkhoff** [comment]
CC @dthittner

**Comment by larsbrinkhoff** [comment]
I'm now adding a ACC LH/DH IMP interface to SIMH.  This will provide TCP/IP networking for KS10 ITS.

**Comment by larsbrinkhoff** [comment]
@atsampson is working on a DEUNA driver for ITS.

**Comment by markpizz** [comment]
Hmm... 
>> My memory about the unibus ethernet adapters is fuzzier. IIRC we ran into serious problems in the end with the 18/16 bit unibus data formatting and byte swapping issues, and decided that performance would be not worth it.

This problem will still exist without regard to whether it is a NI1010 or a DEUNA.  I know nothing about programming the NI1010, but I can only imagine the strangeness involved in the DEUNA walking buffer descriptor lists in the memory of a 36 bit machine and what/how any sense can be made of that...

**Comment by hrlzm** [comment]
Anyone interested can take a look at my DEUNA/DELUA driver for Tops-10.  The code
can be found at:

  https://www.pdc.kth.se/~bygg/tops/unxser.mac

The code works both on SIMH and on real HW, even if said HW has gone west for
the time beeing.

The DEUNA just sees a block of 16-bit words at specified addresses, as long as they
contain the right bits it need not worry about the fact that those words live in some
36-bit wide memory somewhere.

--Johnny

**Comment by atsampson** [comment]
The current status of my DEUNA hack is on the [ats/ethernet](https://github.com/PDP-10/its/tree/ats/ethernet) branch - although you'd probably do better to wait until I've cleaned it up a bit (read: rewritten most of it now I know how I'd really like to do it). It seems to work happily now, at least for incoming telnet.

The DEUNA descriptors actually lay out quite nicely in 36-bit memory, as Johnny says - they're easier to work with in un-BLTUB-ed form. So my code builds the descriptors directly in the shared memory, and BLTUB/BLTBUs the packet contents to/from ITS's buffers as appropriate.

**Comment by larsbrinkhoff** [comment]
Great news, @atsampson!

**Comment by larsbrinkhoff** [comment]
Please remember to update the KS10 section in doc/peripherals.md.

**Comment by hrlzm** [comment]
I had a quick look at the code, and as far as I can tell it will not work on a real KS10.

The shortcut done by the UBA, namely a write access to the leftmost byte in a word
will not be done with a read/modify/write cycle, and therefore overwrite the right half.
The DEUNA will do such a write when updating the OWN bit in the ring entry.  This can
be mitigated by tweaking the ring layout in memory, see my Tops-10 driver for details.

The SIMH deuna implementation does not do this, at least my initial implementation
of it did not.

--Johnny

**Comment by atsampson** [comment]
Cheers, Johnny! I'll look at using your workaround - it should be straightforward enough.

**Comment by larsbrinkhoff** [comment]
ITS running on SIMH should be able to use the CH11 device.  I haven't tested this, and it's not high on my priority list.
